README.md
license: SEE LICENSE IN LICENSE.md [MUST NOT BE REMOVED]

### Mock API Server
Create your own mock server with API documentation services such as Swagger.

##### Currently supports
- Swagger

##### Requirements
- Python 3.x
- pip (based on Python 3.x)

##### Steps to use
1. Clone the repository.
2. `cd` to your project.
3. [You may skip this step] Create virtual environment.
4. Install required libraries using `pip install -r requirements.txt`
4. Download your `swagger.json` file from Swagger UI.
5. Use the following to serve your file `python app.py -s /path/to/swagger.json`
##### Extended use
```
usage: app.py [-h] -s SOURCE [-p PORT] [-t {swagger}] [-sc STATUS_CODE] [-r RANDOM_SEED] [-l LIST_SIZE [LIST_SIZE ...]] [-m [META]]
              [-d DEFAULT_VALUE [DEFAULT_VALUE ...]]

Options for mock responses

Required and optional arguments:
  -h, --help            show this help message and exit

  -s SOURCE, --source SOURCE
                        (Required) API reference source file path.
  
  -p PORT, --port PORT  (Optional,default=5000) Port number the app runs on.

  -t {swagger}, --type {swagger}
                        (Optional,default='swagger') Type of API reference. Currently only supports Swagger.

  -sc STATUS_CODE, --status_code STATUS_CODE
                        (Optional,default=200) Generates responses with status code provided.

  -r RANDOM_SEED, --random_seed RANDOM_SEED
                        (Optional) Generates random responses based on seed value.

  -l LIST_SIZE [LIST_SIZE ...], --list_size LIST_SIZE [LIST_SIZE ...]
                        (Optional,default=[2]) Sets default size of list in response body.

  -m [META], --meta [META]
                        (Optional,default=[true]) Shows the datatype of all parameters
  
  -d DEFAULT_VALUE [DEFAULT_VALUE ...], --default_value DEFAULT_VALUE [DEFAULT_VALUE ...]
                        (Optional) Sets default values in response body. Format key=value.
```

If you would like to donate or know more contact: `bikcrum@gmail.com`

Thank you! Happy coding!

CHANGELOG.md
license: SEE LICENSE IN LICENSE.md [MUST NOT BE REMOVED]

#### CHANGELOG.md

#### 1.0.0
Feb 26, 2020
##### What's new
- Initial release

MockApi.py
# license: SEE LICENSE IN LICENSE.md [MUST NOT BE REMOVED]

from Preference import Preference
from RequestBuilder import SwaggerRequestBuilder
from RequestModel import Request


class BaseMockApi:
    def __init__(self, preference):
        self.requests: [Request] = []
        if preference is None:
            raise Exception('Preference cannot be null')
        self.preference = preference

    def get_routes(self) -> [tuple]:
        return [{'rule': request.request_url, 'method': request.method} for request in self.requests]

    def process_request(self,
                        user_request,
                        params):

        for request in self.requests:
            validated_data = request.validate(url=user_request.url_rule.rule,
                                              method=user_request.method,
                                              req_body=None,
                                              params=params,
                                              pref=self.preference)
            if validated_data is not None:
                return validated_data

        raise Exception('Cannot find request to handle the url')

    def show(self):
        for request in self.requests:
            print(f"-> {request.method} {request.request_url}")


class MockApi(BaseMockApi):
    def __init__(self, api_type, preference: Preference):
        super().__init__(preference)
        self.api_type = api_type

    def build_requests(self, path):
        if self.api_type == 'swagger':
            self.requests = SwaggerRequestBuilder(path).build()
            return

        raise Exception('Cannot build request for unknown type. Currenly only swagger is supported')


RequestModel.py
# license: SEE LICENSE IN LICENSE.md [MUST NOT BE REMOVED]

import json

from Preference import Preference
from Randomizer import Randomizer


class Request:
    def __init__(self):
        self.request_url = None
        self.method = None
        self.parameters = None
        self.request_schema = None
        self.responses: [Response] = None

    def validate(self, url,
                 method,
                 req_body,
                 params,
                 pref: Preference):

        if self.request_url == url and method.lower() == self.method.lower():
            for response in self.responses:
                if response.status_code == str(pref.status_code):
                    return json.dumps(response.get_response_body(pref=pref))

        return None

    def __str__(self):
        return '\n'.join(['Request',
                          f'-> request_url:{self.request_url}',
                          f'-> method:{self.method}',
                          f'-> parameters:{self.parameters}',
                          f'-> request_body:{self.request_schema}',
                          f'-> response:{[str(response) for response in self.responses]}'])


class Response:
    def __init__(self):
        self.status_code = None
        self.response_schema = None

    @staticmethod
    def _get_enum_list(values):
        return f"enum({','.join(map(str, values))})"

    def _data_from_schema(self,
                          prop_name,
                          schema,
                          randomizer: Randomizer,
                          pref: Preference):
        if type(schema) is dict:
            # leaf level
            if schema.get('type') == 'string':
                if prop_name in pref.default_value:
                    return str(pref.default_value[prop_name]), 'string'

                if type(schema.get('enum')) is list:
                    return randomizer.get_random_from_list(schema.get('enum')), \
                           'string.' + self._get_enum_list(schema.get('enum'))

                return randomizer.get_random_string(), 'string'

            if schema.get('type') == 'integer':
                if prop_name in pref.default_value:
                    return int(pref.default_value[prop_name]), 'integer'

                if type(schema.get('enum')) is list:
                    return randomizer.get_random_from_list(schema.get('enum')), \
                           'integer.' + self._get_enum_list(schema.get('enum'))

                return randomizer.get_random_integer(), 'integer'

            if schema.get('type') == 'number':
                if prop_name in pref.default_value:
                    return float(pref.default_value[prop_name]), 'number'

                if type(schema.get('enum')) is list:
                    return randomizer.get_random_from_list(schema.get('enum')), \
                           'number.' + self._get_enum_list(schema.get('enum'))

                return randomizer.get_random_number(), 'number'

            if schema.get('type') == 'boolean':
                if prop_name in pref.default_value:
                    return bool(pref.default_value[prop_name]), 'boolean'

                if type(schema.get('enum')) is list:
                    return randomizer.get_random_from_list(schema.get('enum')), \
                           'boolean.' + self._get_enum_list(schema.get('enum'))

                return randomizer.get_random_boolean(), 'boolean'

            # non-leaf level
            if schema.get('type') == 'object':
                new_dict = {}
                for prop_name in schema['properties']:
                    data = self._data_from_schema(prop_name, schema['properties'][prop_name], randomizer, pref)
                    if type(data) is tuple:
                        new_dict[prop_name] = data[0]
                        if pref.meta:
                            new_dict['@' + prop_name] = data[1]
                    else:
                        new_dict[prop_name] = data

                return new_dict

            if schema.get('type') == 'array':
                size = randomizer.get_random_from_list(pref.list_size)

                items = [self._data_from_schema('items', schema['items'], randomizer, pref) for _ in
                         range(size)]

                return items

        return None

    def get_response_body(self, pref: Preference):

        randomizer = Randomizer()
        if pref.random_seed is not None:
            randomizer.set_random_seed(pref.random_seed)

        return self._data_from_schema(None,
                                      schema=self.response_schema,
                                      randomizer=randomizer,
                                      pref=pref)

    def __str__(self):
        return '\n'.join(['Response',
                          f'-> status_code:{self.status_code}',
                          f'-> response_body:{self.response_schema}'])


RequestBuilder.py
# license: SEE LICENSE IN LICENSE.md [MUST NOT BE REMOVED]

import json
import os

from RequestModel import Response, Request


class BaseRequestBuilder:
    def __init__(self, filepath):
        self.filepath = filepath
        self.buffer = open(filepath, 'r')
        os.makedirs('generated', exist_ok=True)
        self.generated = open(os.path.join('generated', os.path.basename(filepath)), 'w')


class SwaggerRequestBuilder(BaseRequestBuilder):
    def __init__(self, filepath):
        super().__init__(filepath)

        data = json.load(self.buffer)
        self.data = self.build_reference(data)
        self.generated.write(json.dumps(self.data, indent=4))

    def build_reference(self, data):
        return self._build_reference(data, data, [])

    def _build_reference(self, _data, data, references):
        if type(data) is dict:
            new_dict = {}
            for key in data:
                if key == '$ref':
                    ref = self._find_schema(_data, data.get(key))

                    # Avoiding cyclic dependency
                    if data.get(key) in references:
                        return f"Cyclic dependency error: Cannot reference {data.get(key)} after {' -> '.join(references)}"

                    new_dict = self._build_reference(_data, ref, references + [data.get(key)])
                    return new_dict
                else:
                    new_dict[key] = self._build_reference(_data, data.get(key), references)
            return new_dict

        if type(data) is list:
            new_list = []
            for val in data:
                new_list.append(self._build_reference(_data, val, references))
            return new_list

        return data

    @staticmethod
    def _find_schema(_data, ref):
        path = ref[2:].split('/')
        ref = _data
        for p in path:
            ref = ref[p]

        return ref

    def build(self) -> [Request]:
        requests = []

        for path, child in self.data['paths'].items():
            for method, _child in child.items():
                request = Request()
                request.method = method

                if 'parameters' in _child:
                    request.parameters = [parameter['name'] for parameter in _child['parameters']]
                    for param in request.parameters:
                        path = path.replace('{' + param + '}', '<' + param + '>')
                request.request_url = path

                if 'responses' in _child:
                    request.responses = []
                    for status_code in _child['responses']:
                        response = Response()
                        response.status_code = status_code

                        if 'content' in _child['responses'][status_code]:
                            for content_type in _child['responses'][status_code]['content']:
                                response.response_schema = _child['responses'][status_code]['content'][content_type][
                                    'schema']
                                break
                        request.responses.append(response)

                if 'requestBody' in _child:
                    if 'content' in _child['requestBody']:
                        for content_type in _child['requestBody']['content']:
                            request.request_schema = _child['requestBody']['content'][content_type]['schema']
                            break

                requests.append(request)

        return requests


app.py
# license: SEE LICENSE IN LICENSE.md [MUST NOT BE REMOVED]

import argparse

from flask import Flask, Response
from flask import request

from MockApi import MockApi
from Preference import Preference


def build_arguments():
    parser = argparse.ArgumentParser(description='Options for mock responses')
    parser.add_argument('-s',
                        '--source',
                        required=True,
                        help='(Required) API reference source file path.')
    parser.add_argument('-p',
                        '--port',
                        type=str,
                        help='(Optional,default=5000) Port number the app runs on.',
                        default=5000)
    parser.add_argument('-t',
                        '--type',
                        type=str,
                        help='(Optional,default=\'swagger\') Type of API reference. Currently only supports Swagger.',
                        choices=['swagger'],
                        default='swagger')
    parser.add_argument('-sc',
                        '--status_code',
                        type=int,
                        help='(Optional,default=200) Generates responses with status code provided',
                        default=200)
    parser.add_argument('-r',
                        '--random_seed',
                        type=int,
                        help='(Optional) Generates random responses based on seed value')
    parser.add_argument('-l',
                        '--list_size',
                        nargs='+',
                        type=int,
                        help='(Optional,default=[2]) Sets default size of list in response body',
                        default=[2])

    def str2bool(v):
        if isinstance(v, bool):
            return v
        if v.lower() in ('yes', 'true', 't', 'y', '1'):
            return True
        elif v.lower() in ('no', 'false', 'f', 'n', '0'):
            return False
        else:
            raise argparse.ArgumentTypeError('Boolean value expected.')

    parser.add_argument('-m',
                        '--meta',
                        type=str2bool,
                        nargs='?',
                        const=True,
                        help='(Optional,default=[true]) Shows the datatype of all parameters',
                        default=True)
    parser.add_argument('-d',
                        '--default_value',
                        nargs='+',
                        help='(Optional) Sets default values in response body. Format key=value')

    return parser.parse_args()


def process_request(**kwargs):
    return Response(mp.process_request(user_request=request,
                                       params=kwargs),
                    mimetype='application/json',
                    status=args.status_code if args.status_code is not None else 200)


if __name__ == '__main__':
    args = build_arguments()
    print(args)

    pref = Preference(
        status_code=args.status_code,
        random_seed=args.random_seed,
        default_value=dict([value.split('=') for value in args.default_value])
        if args.default_value is not None else [],
        list_size=args.list_size,
        meta=args.meta
    )

    mp = MockApi(api_type=args.type, preference=pref)
    mp.build_requests(args.source)

    app = Flask(__name__)

    for route in mp.get_routes():
        view_func = app.route(route['rule'], methods=[route['method']])(process_request)

    print('Active endpoints')
    mp.show()

    app.run(port=args.port)
